### PR TITLE
Hide the implementation of `scheduler.Config`

### PR DESCRIPTION
**Pull request by tossmilestone**
**What this PR does / why we need it**:
Fix the following `TODO`: 
https://github.com/kubernetes/kubernetes/blob/317853c90c674920bfbbdac54fe66092ddc9f15f/pkg/scheduler/scheduler.go#L100
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
None

**Special notes for your reviewer**:

**Release note**:

```release-note
None
```

### CODE REVIEW

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -97,46 +97,88 @@ type Configurator interface {
...
+		nextPod:             nextPod,
+		error:               error,
+		Recorder:            recorder,
+		StopEverything:      stopEverything,
+		volumeBinder:        volumeBinder,
+	}
+}
+
+// Algorithm returns the ScheduleAlgorithm object of the config.
+func (c *Config) Algorithm() algorithm.ScheduleAlgorithm {
```
Comment (bsalamat) [inline pkg/scheduler/scheduler.go]: Providing getters for "Algorithm" and "SchedulerCache" and not provding getters for "StopEverything" and "Recorder" and leaving the latters publicly accessible make the code inconsistent. Either remove the getters and make the other two publicly accessible or provide getters for all of them.

Comment (tossmilestone) [reply pkg/scheduler/scheduler.go]: Agreed. I will add getters for `StopEverything` and `Recorder`

**Comment by tossmilestone** [comment]
@bsalamat Updated, PTAL :smile:

**Review by bsalamat — COMMENTED** [review]
LGTM, just a minor comment.

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -97,46 +97,108 @@ type Configurator interface {
...
+func (c *Config) SetRecorder(recorder record.EventRecorder) {
+	c.recorder = recorder
+}
+
+// StopEverything returns the stopEverything channel of the config.
+func (c *Config) StopEverything() chan struct{} {
+	return c.stopEverything
+}
+
+// SetStopEverything sets the stopEverything channel of the config.
```
[the remark this answers is not in the document]
Comment (tossmilestone) [reply pkg/scheduler/scheduler.go]: Updated, PTAL @bsalamat

**Comment by bsalamat** [comment]
LGMT
You may squash the commits. Thanks!

**Review by resouer — COMMENTED** [review]
/lgtm

Really sorry for late review :)

**Comment by tossmilestone** [comment]
/assign @timothysc
Needs your approval for `test` changes, thanks! :smile:

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -97,46 +97,109 @@ type Configurator interface {
 }
 
 // Config is an implementation of the Scheduler's configured input data.
-// TODO over time we should make this struct a hidden implementation detail of the scheduler.
 type Config struct {
```
Comment (liggitt) [inline pkg/scheduler/scheduler.go]: Do we know there are no consumers driving the scheduler via this? /cc @aveshagarwal

Comment (liggitt) [reply pkg/scheduler/scheduler.go]: Is a config struct with private members like this consistent with how other components are driven?

Comment (tossmilestone) [reply pkg/scheduler/scheduler.go]: For `kube-apiserver`, `kubelet` and `kube-controller-manager`, all of them expose the config's members. Now I doubt if the `TODO` is reasonable or there are special purposes of it.

Comment (tossmilestone) [reply pkg/scheduler/scheduler.go]: @liggit Please give some suggestions on if the `TODO` is necessary? Thanks!
:smiley:

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/59724)

File: pkg/scheduler/scheduler.go (hunk 3 above)
Comment (liggitt) [reply pkg/scheduler/scheduler.go]: I would probably remove the TODO and wait to move scheduler in this direction until it's clear all the components are doing so

Comment (tossmilestone) [reply pkg/scheduler/scheduler.go]: Agreed. I will just remove the `TODO` then.

**Comment by tossmilestone** [comment]
Discussed with @liggitt I decided to abandoned the previous changes and just remove the `TODO`.
PTAL @bsalamat

**Comment by tossmilestone** [comment]
@bsalamat because hide the implementation of config will break the consistentency with the other components that exposes their config's members, I think the configuration maybe need a unified design to avoid the inconsistency.
Also a reason quoted from @liggitt :
>I would probably remove the TODO and wait to move scheduler in this direction until it's clear all the components are doing so

**Comment by bsalamat** [comment]
@tossmilestone other components are not going to change on their own and deleting this TODO is not going to help here. I would close this PR and leave the TODO for now.

**Comment by tossmilestone** [comment]
OK, agree @bsalamat

**Pull request closed without merging**